app/models/reduction_fetcher.rb

```diff
@@ -1,49 +1,51 @@
-class ReductionFetcher
-  attr_accessor :topic
-  def initialize(filter)
-    @filter = filter
-    @subject_reductions = SubjectReduction.where(filter.except(:user_id))
-    @user_reductions = UserReduction.where(filter.except(:subject_id))
-
-    @topic = :reduce_by_subject
+class ReductionFetcher < FetcherBase
+  attr_reader :loaded
+
+  def initialize(query)
+    super(query)
+    @loaded = false
+
+    @subject_reductions = SubjectReduction.where(query.except(:user_id))
+    @user_reductions = UserReduction.where(query.except(:subject_id))
   end
 
   def load!
+    return self if loaded
+
     @subject_reductions.load
     @user_reductions.load
-  end
+    @loaded = true
 
-  def for!(topic)
-    @topic = topic.to_sym
     self
   end
 
-  def retrieve(**kwargs)
-    selector = @filter.merge(kwargs)
-    return @subject_reductions.where(selector.except(:user_id)).first_or_initialize if reduce_by_subject?
-    return @user_reductions.where(selector.except(:subject_id)).first_or_initialize if reduce_by_user?
-  end
-
-  def retrieve_in_place(**kwargs)
-    selector = @filter.merge(kwargs)
-    return locate_in_place(selector.except(:user_id), @subject_reductions, SubjectReduction) if reduce_by_subject?
-    return locate_in_place(selector.except(:subject_id), @user_reductions, UserReduction) if reduce_by_user?
-  end
+  def search(**kwargs)
+    selector = query.merge(kwargs)
+    selector.delete(:user_id) if fetch_by_subject?
+    selector.delete(:subject_id) if fetch_by_user?
 
-  def reduce_by_user?
-    @topic == :reduce_by_user
+    if loaded
+      locate_in_place(selector)
+    else
+      source_relation.where(selector).to_a
```

Comment:
i think i did this for testing purposes and forgot to take it back out.

Reply:
actually no, this is because the `Reducer#process` method now expects to receive an array of reduction objects instead of an AR relation, so it's kind of necessary